runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -282,6 +298,84 @@ public void close() throws Exception {
     }
   }
 
+  /** A {@link StageBundleFactory} that supports environment expiration. */
+  private class SmarterStageBundleFactory implements StageBundleFactory {
```

Comment:
Maybe rename to `ExpirableStageBundleFactory`?

Reply:
Also, I wonder if we should simply integrate this into the existing SimpleStageBundleFactory and default to never expire there.

Reply:
Integrated.